### PR TITLE
Loader refactor

### PR DESCRIPTION
**Pull request by privat**
A big refactorization of the loader

This include cleanup, renaming and documentation of the existing code.

The major change is the removal of `ModulePath` used to designate an identified module that is not yet loaded. This caused major issues since once the module was loaded, the MModule object has to be used instead.
Now, genuine `MModule` are used to also represent modules identified in the file system but not yet loaded.
This simplify the code of the loader, nitls and nitcatalog.

By the way, the option `-d` (and `-M`) of nitls that was broken by the previous changes in the loader is now fixed.

Another change is the introduction of `scan_full` that is a high-level method to be used by clients like nitls or nitcatalog that just need to identify packages, groups and modules in the filesystem without loading them.

By the way, nitcalatog works on directories of projects now. eg. `nitcatalog ../lib)

Note, because this PR introduces a lot of interrelated changes, It was hard to come up with nice commits. The alternative was to have a single big commit.

### CODE REVIEW

File: src/loader.nit [hunk 1]
```diff
@@ -15,6 +15,26 @@
...
+# Scanning is done to enable the searching of modules in projects.
+# See `scan_group` and `scan_full`.
+#
+# Parse: means loading the AST and associating with the model entity.
+# See `MModule::parse`.
+#
+# Import: means recursively load modules imported by a module.
+# See `build_module_importation`.
+#
+# Load: means doing the full sequences: identify, parse, import.
```
Comment (xymus) [inline src/loader.nit]: All the `means ...ing` do not add much information, they could be removed for a cleaner looking doc. :wink: 

ex: `Identify: create an* empty model entity associated to a name or a file path.`

**Comment by xymus** [comment]
+1

File: src/loader.nit [hunk 2]
```diff
@@ -92,52 +110,18 @@ redef class ModelBuilder
...
+	# Identify a bunch of modules and groups.
 	#
-	# * a module, then is it parser and returned.
-	# * a group then recursively all its modules are parsed.
-	# * a directory of packages then all the modules of all packages are parsed.
-	# * else an error is displayed.
-	#
-	# See `parse` for details.
-	fun parse_full(names: Sequence[String]): Array[MModule]
+	# This does the same that `parse_full` but does only the identification (cf. `identify_module`)
```
Comment (R4PaSs) [inline src/loader.nit]: :s/that/as

File: src/loader.nit [hunk 3]
```diff
@@ -169,12 +149,41 @@ redef class ModelBuilder
...
+	# Load a bunch of modules and groups.
+	#
+	# Each name can be:
+	#
+	# * a path to a module, a group or a directory of packages.
+	# * a short name of a module or a group that are looked in the `paths` (-I)
+	#
+	# Then, for each entry, if it is:
+	#
+	# * a module, then is it parser and returned.
```
Comment (R4PaSs) [inline src/loader.nit]: :s/parser/parsed

**Comment by privat** [comment]
reroll: pass tests and apply comments

**Pull request merged**